#Méthodologie CRISP :




##Etape 1 : Compréhension de métier (Business understanding)


##Etape 2 : Compréhension de données (Data understanding)


In [ ]:
import sklearn.datasewts

In [ ]:
from google.colab import drive
drive.mount('/content/drive' , force_remount=True)

Mounted at /content/drive


In [ ]:
container_path='/content/drive/MyDrive/bbc'
data_dir = sklearn.datasets.load_files(container_path, description=None, categories=None, 
                        load_content=True, shuffle=True, encoding=None, decode_error='strict', random_state=0)

In [ ]:
y=data_dir.target
X=data_dir.data

In [ ]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,
                                               random_state=42, 
                                               test_size=0.3,
                                               stratify=None)

##Etape 3 : Préparation des données (Data preprocessing)


3.1.Transformation BOW :

3.1.1. Application du modèle BOW par défaut

In [ ]:
def simple_tokenizer(s):
  return s.split()

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer

In [ ]:
bow=CountVectorizer(tokenizer=simple_tokenizer)
bow.fit(X_train)

/usr/local/lib/python3.6/dist-packages/sklearn/feature_extraction/text.py:507: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn("The parameter 'token_pattern' will not be used"


CountVectorizer(analyzer='word', binary=False, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=None, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words=None,
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=<function simple_tokenizer at 0x7fc2a74f4ea0>,
                vocabulary=None)

In [ ]:
vocab=bow.get_feature_names()
vocab

['!',
 '"',
 '"\'i',
 '"\'we',
 '"\'you',
 '"...it',
 '"1983',
 '"2+2=5".',
 '"2004',
 '"2004\'s',
 '"2005',
 '"21st',
 '"27%',
 '"3g',
 '"4th',
 '"50\'s',
 '"50-50"',
 '"50/50".',
 '"[alastair]',
 '"[but]',
 '"[godzilla\'s]',
 '"[i\'m]',
 '"[it]',
 '"[oil',
 '"[phone-tap',
 '"[the',
 '"a',
 '"a"',
 '"abandonment',
 '"abhorrent',
 '"ability',
 '"abnormal',
 '"about',
 '"above',
 '"absolute',
 '"absolutely',
 '"absolutely"',
 '"abstentions',
 '"absurd',
 '"absurd"',
 '"abuse',
 '"abusive',
 '"acceptance',
 '"accepting',
 '"access',
 '"accordingly,',
 '"acknowledged',
 '"acquiescent"',
 '"acting',
 '"action',
 '"adam',
 '"add-ons"',
 '"added',
 '"addictive',
 '"adherence',
 '"adjusted"',
 '"adults',
 '"adults"',
 '"advertisers',
 '"affect"',
 '"affordable"',
 '"affordable".',
 '"afraid',
 '"africa',
 '"after',
 '"against',
 '"agents',
 '"aggregator".',
 '"aggressive',
 '"aggressive"',
 '"aggressively',
 '"agnostic"',
 '"agriculture',
 '"agroflora',
 '"ah',
 '"aims',
 '"airbus',
 '"airy',

In [ ]:
X1=bow.transform(X_train)
X1

<1557x50742 sparse matrix of type '<class 'numpy.int64'>'
	with 347635 stored elements in Compressed Sparse Row format>

3.1.2. Etude de la tokenisation au niveau du modèle BOW



In [ ]:
import nltk
text="Feet cat's wolves ours you talked"
tokenizer = nltk.tokenize.TreebankWordTokenizer()
tokens= tokenizer.tokenize(text)
tokens
tokens_mini=[]
for w in tokens:
  tokens_mini.append(w.lower())
tokens_mini


['feet', 'cat', "'s", 'wolves', 'ours', 'you', 'talked']

In [ ]:
text="Feet cat's wolves ours, you talked"
tokenizer = nltk.tokenize.WhitespaceTokenizer()
tokens= tokenizer.tokenize(text)
tokens

['Feet', "cat's", 'wolves', 'ours,', 'you', 'talked']

In [ ]:
text="Feet cat's wolves ours, you talked"
tokenizer = nltk.tokenize.WordPunctTokenizer()
tokens= tokenizer.tokenize(text)
tokens

['Feet', 'cat', "'", 's', 'wolves', 'ours', ',', 'you', 'talked']

In [ ]:
text="Feet cat's wolves ours, you talked"
tokenizer = nltk.tokenize.TreebankWordTokenizer()
tokens= tokenizer.tokenize(text)
tokens

['Feet', 'cat', "'s", 'wolves', 'ours', ',', 'you', 'talked']

In [ ]:
text="Feet cat's wolves ours, you talked"
tokenizer = nltk.tokenize.RegexpTokenizer('\w+|\$[\d\.]+|\S+')
tokens= tokenizer.tokenize(text)
tokens

['Feet', 'cat', "'s", 'wolves', 'ours', ',', 'you', 'talked']

In [ ]:
text="Feet cat's wolves ours, you talked"
stemmer = nltk.stem.PorterStemmer()
stemmer
" ".join(stemmer.stem(token) for token in tokens)

"feet cat 's wolv our , you talk"

In [ ]:
from nltk.stem import WordNetLemmatizer 
import nltk
nltk.download('wordnet')
lemmatizer = WordNetLemmatizer() 

" ".join(lemmatizer.lemmatize(token) for token in tokens)

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Unzipping corpora/wordnet.zip.


"Feet cat 's wolf ours , you talked"

In [ ]:
from nltk.corpus import stopwords
nltk.download('stopwords')

print(stopwords.words('english'))
for w in tokens_mini[:]:
  if ((w in stopwords.words('english'))==False) :
    tokens_mini.remove(w)


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'ea

In [ ]:
stemmer = nltk.stem.PorterStemmer()
" ".join(stemmer.stem(token) for token in tokens)

"feet cat 's wolv our , you talk"

In [ ]:
from nltk.corpus import stopwords
import nltk

def segmenter_texte(texte):
  tokenizer = nltk.tokenize.TreebankWordTokenizer()
  tokens= tokenizer.tokenize(texte)
  tokens_mini=[]
  for w in tokens:
    tokens_mini.append(w.lower())
    #  nltk.download('stopwords')
  for w in tokens_mini[:]:
    if w in stopwords.words('english') :
      tokens_mini.remove(w)
  tokens=tokens_mini
  stemmer = nltk.stem.PorterStemmer()
  ch=" ".join(stemmer.stem(token) for token in tokens)
  return ch.split()


In [ ]:
segmenter_texte("our your sofiane hello world you he him his are perfect")

['sofian', 'hello', 'world', 'perfect']

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
bow2=CountVectorizer(tokenizer=segmenter_texte)
bow2.fit(X_train)

/usr/local/lib/python3.6/dist-packages/sklearn/feature_extraction/text.py:507: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn("The parameter 'token_pattern' will not be used"


CountVectorizer(analyzer='word', binary=False, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=None, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words=None,
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=<function segmenter_texte at 0x7fc29dcfc6a8>,
                vocabulary=None)

In [ ]:
  vocabulaire=bow2.get_feature_names()
  vocabulaire

['!',
 '#',
 '$',
 '%',
 '&',
 "'",
 "''",
 "''he",
 "''i",
 "''person",
 "''poach",
 "''robbi",
 "'.",
 "'15",
 "'2004",
 "'771",
 "'83",
 "'93",
 "'94",
 "'a",
 "'a'.",
 "'absolut",
 "'afflict",
 "'ahead",
 "'all",
 "'alleg",
 "'alway",
 "'anfield",
 "'angadia",
 "'angel",
 "'anti-semitic'",
 "'artist",
 "'ass-sav",
 "'at",
 "'audienc",
 "'automat",
 "'awar",
 "'back",
 "'bad",
 "'bb+'.",
 "'bbb-'.",
 "'be",
 "'befor",
 "'best",
 "'bi",
 "'bid",
 "'big",
 "'black",
 "'blackburn",
 "'blatant",
 "'blip-pop'",
 "'block",
 "'blog",
 "'blogospher",
 "'bold",
 "'bot",
 "'brainwav",
 "'bribe",
 "'bring",
 "'budget",
 "'bureaucraci",
 "'can",
 "'cannibals'",
 "'cantona",
 "'card-not-pres",
 "'career",
 "'ceremoni",
 "'certif",
 "'chang",
 "'charad",
 "'chase",
 "'cheat",
 "'chip",
 "'christma",
 "'clean",
 "'clear",
 "'cold",
 "'comeback",
 "'comput",
 "'confus",
 "'copycat'",
 "'could",
 "'cozza",
 "'crunk'.",
 "'cut",
 "'d",
 "'damag",
 "'danc",
 "'debat",
 "'declin",
 "'defi",
 "'delay",


In [ ]:
X_train_bow=bow2.transform(X_train)
X_train_bow

<1557x28238 sparse matrix of type '<class 'numpy.int64'>'
	with 258452 stored elements in Compressed Sparse Row format>

On comparent la resultat du bow avec une separation normal 347635 to 258452

3.1.3. Etude des mots/tokens de vocabulaire au niveau du modèle BOW


In [ ]:
bow=CountVectorizer(tokenizer=segmenter_texte ,ngram_range=(1,1))
bow.fit(X_train)
vocabulaire=bow.get_feature_names()
vocabulaire
len(vocabulaire)

/usr/local/lib/python3.6/dist-packages/sklearn/feature_extraction/text.py:507: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn("The parameter 'token_pattern' will not be used"


28238

In [ ]:
bow=CountVectorizer(tokenizer=segmenter_texte ,ngram_range=(2,2))
bow.fit(X_train)
vocabulaire=bow.get_feature_names()
vocabulaire

/usr/local/lib/python3.6/dist-packages/sklearn/feature_extraction/text.py:507: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn("The parameter 'token_pattern' will not be used"


['! !',
 "! ''",
 "! 'd",
 "! 'm",
 "! 's",
 "! 've",
 '! (',
 '! )',
 '! ,',
 '! absolut',
 '! accept',
 '! agre',
 '! along',
 '! also',
 '! amus',
 '! apple..',
 '! arrest',
 '! back',
 '! base',
 '! believ',
 '! best',
 '! black',
 '! bring',
 '! ca',
 '! calm',
 '! colleagu',
 '! controversi',
 '! day',
 '! editor',
 '! entertainment.',
 '! everi',
 '! feel',
 '! film',
 '! first',
 '! get',
 '! good',
 '! got',
 '! govern',
 '! graduat',
 '! great',
 '! hate',
 '! howev',
 '! impress',
 '! interest',
 '! job',
 '! join',
 '! joss',
 '! keanu',
 '! keep',
 '! last',
 '! leav',
 '! life',
 '! list',
 '! listen',
 '! lol',
 '! love',
 '! magazin',
 '! must',
 "! n't",
 '! noth',
 '! obvious',
 '! ok',
 '! one',
 '! onlin',
 '! part',
 '! partner',
 '! pave',
 '! peopl',
 '! perhap',
 '! play',
 '! playwright',
 '! poll',
 '! pray',
 '! predict',
 '! re-gurgit',
 '! real',
 '! recent',
 '! rememb',
 '! right',
 '! robbi',
 '! ruin',
 '! sad',
 '! said',
 '! scandinavian',
 '! seen',


In [ ]:
bow=CountVectorizer(tokenizer=segmenter_texte ,ngram_range=(1,2))
bow.fit(X_train)
vocabulaire=bow.get_feature_names()
vocabulaire

/usr/local/lib/python3.6/dist-packages/sklearn/feature_extraction/text.py:507: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn("The parameter 'token_pattern' will not be used"


['!',
 '! !',
 "! ''",
 "! 'd",
 "! 'm",
 "! 's",
 "! 've",
 '! (',
 '! )',
 '! ,',
 '! absolut',
 '! accept',
 '! agre',
 '! along',
 '! also',
 '! amus',
 '! apple..',
 '! arrest',
 '! back',
 '! base',
 '! believ',
 '! best',
 '! black',
 '! bring',
 '! ca',
 '! calm',
 '! colleagu',
 '! controversi',
 '! day',
 '! editor',
 '! entertainment.',
 '! everi',
 '! feel',
 '! film',
 '! first',
 '! get',
 '! good',
 '! got',
 '! govern',
 '! graduat',
 '! great',
 '! hate',
 '! howev',
 '! impress',
 '! interest',
 '! job',
 '! join',
 '! joss',
 '! keanu',
 '! keep',
 '! last',
 '! leav',
 '! life',
 '! list',
 '! listen',
 '! lol',
 '! love',
 '! magazin',
 '! must',
 "! n't",
 '! noth',
 '! obvious',
 '! ok',
 '! one',
 '! onlin',
 '! part',
 '! partner',
 '! pave',
 '! peopl',
 '! perhap',
 '! play',
 '! playwright',
 '! poll',
 '! pray',
 '! predict',
 '! re-gurgit',
 '! real',
 '! recent',
 '! rememb',
 '! right',
 '! robbi',
 '! ruin',
 '! sad',
 '! said',
 '! scandinavian',
 '! s

In [ ]:
bow=CountVectorizer(tokenizer=segmenter_texte ,ngram_range=(1,1),min_df=2, max_df=5)
bow.fit(X_train)
vocabulaire=bow.get_feature_names()
vocabulaire


/usr/local/lib/python3.6/dist-packages/sklearn/feature_extraction/text.py:507: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn("The parameter 'token_pattern' will not be used"


["'83",
 "'a'.",
 "'absolut",
 "'best",
 "'bot",
 "'brainwav",
 "'cheat",
 "'confus",
 "'do",
 "'easi",
 "'em",
 "'eu",
 "'face",
 "'fail",
 "'found",
 "'gadget",
 "'given",
 "'goodby",
 "'govern",
 "'greatest",
 "'grey",
 "'hear",
 "'here",
 "'hit",
 "'if",
 "'it",
 "'law",
 "'lib",
 "'lies'",
 "'look",
 "'lose",
 "'may",
 "'me",
 "'more",
 "'must",
 "'must-haves'",
 "'n",
 "'new",
 "'no",
 "'not",
 "'pocket",
 "'reject",
 "'set",
 "'should",
 "'super",
 "'take",
 "'templat",
 "'the",
 "'thi",
 "'top",
 "'tough",
 "'urban",
 "'verita",
 "'wait",
 "'we",
 "'what",
 "'will",
 "'would",
 "'ye",
 "'you",
 '.aero.',
 '.biz',
 '.com',
 '.info',
 '.museum',
 '.post',
 '.travel',
 '.xxx',
 '0-0',
 '0-2',
 '0.03',
 '0.4',
 '0.5',
 '0.6',
 '0.8',
 '007',
 '0800',
 '0900',
 '1,000.',
 '1,000th',
 '1,080',
 '1,100',
 '1,200',
 '1,300',
 '1,324',
 '1,444',
 '1,500',
 '1,500m',
 '1,600',
 '1,920',
 '1-0.',
 '1-6',
 '1.0',
 '1.07bn',
 '1.1',
 '1.1bn',
 '1.20',
 '1.25',
 '1.26',
 '1.3bn',
 '1.4',
 '1

In [ ]:
len(vocabulaire)

8410

In [ ]:
bow=CountVectorizer(tokenizer=segmenter_texte ,ngram_range=(2,2),min_df=2, max_df=5)
bow.fit(X_train)
vocabulaire=bow.get_feature_names()
vocabulaire


/usr/local/lib/python3.6/dist-packages/sklearn/feature_extraction/text.py:507: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn("The parameter 'token_pattern' will not be used"


["! ''",
 '! ,',
 '! along',
 '! also',
 '! apple..',
 '! back',
 '! everi',
 '! first',
 '! get',
 '! got',
 '! great',
 '! interest',
 '! keep',
 '! must',
 "! n't",
 '! partner',
 '! pave',
 '! recent',
 '! sad',
 '! said',
 '! sinclair',
 '! soni',
 '! stood',
 '! think',
 '! well',
 '# 163',
 '# 191',
 '$ 1',
 '$ 1.07bn',
 '$ 1.1bn',
 '$ 1.20',
 '$ 1.3',
 '$ 1.3bn',
 '$ 1.5bn',
 '$ 1.6bn',
 '$ 1.7bn',
 '$ 1.8',
 '$ 1.8bn',
 '$ 1.8bn.',
 '$ 1.9bn',
 '$ 10.7bn',
 '$ 100bn',
 '$ 10bn',
 '$ 10m',
 '$ 11.5bn',
 '$ 110m',
 '$ 12',
 '$ 120m',
 '$ 125m',
 '$ 13',
 '$ 135m',
 '$ 15',
 '$ 150',
 '$ 150,000',
 '$ 150m',
 '$ 15bn',
 '$ 16.1m',
 '$ 160bn',
 '$ 16bn',
 '$ 16bn.',
 '$ 17m',
 '$ 180bn',
 '$ 184m',
 '$ 19.1m',
 '$ 19bn',
 '$ 2.3bn',
 '$ 2.5bn',
 '$ 2.7bn',
 '$ 2.9bn',
 '$ 20',
 '$ 200,000',
 '$ 200m',
 '$ 206bn',
 '$ 20bn',
 '$ 21m',
 '$ 23bn',
 '$ 24.9m',
 '$ 24m',
 '$ 25.4bn',
 '$ 250',
 '$ 250,000',
 '$ 250m',
 '$ 25m',
 '$ 27.5bn',
 '$ 28.3bn',
 '$ 280bn',
 '$ 3.1bn',
 '$ 3.5b

In [ ]:
len(vocabulaire)

45958

In [ ]:
bow=CountVectorizer(tokenizer=segmenter_texte ,ngram_range=(1,2),min_df=2, max_df=5)
bow.fit(X_train)
vocabulaire=bow.get_feature_names()
vocabulaire


/usr/local/lib/python3.6/dist-packages/sklearn/feature_extraction/text.py:507: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn("The parameter 'token_pattern' will not be used"


["! ''",
 '! ,',
 '! along',
 '! also',
 '! apple..',
 '! back',
 '! everi',
 '! first',
 '! get',
 '! got',
 '! great',
 '! interest',
 '! keep',
 '! must',
 "! n't",
 '! partner',
 '! pave',
 '! recent',
 '! sad',
 '! said',
 '! sinclair',
 '! soni',
 '! stood',
 '! think',
 '! well',
 '# 163',
 '# 191',
 '$ 1',
 '$ 1.07bn',
 '$ 1.1bn',
 '$ 1.20',
 '$ 1.3',
 '$ 1.3bn',
 '$ 1.5bn',
 '$ 1.6bn',
 '$ 1.7bn',
 '$ 1.8',
 '$ 1.8bn',
 '$ 1.8bn.',
 '$ 1.9bn',
 '$ 10.7bn',
 '$ 100bn',
 '$ 10bn',
 '$ 10m',
 '$ 11.5bn',
 '$ 110m',
 '$ 12',
 '$ 120m',
 '$ 125m',
 '$ 13',
 '$ 135m',
 '$ 15',
 '$ 150',
 '$ 150,000',
 '$ 150m',
 '$ 15bn',
 '$ 16.1m',
 '$ 160bn',
 '$ 16bn',
 '$ 16bn.',
 '$ 17m',
 '$ 180bn',
 '$ 184m',
 '$ 19.1m',
 '$ 19bn',
 '$ 2.3bn',
 '$ 2.5bn',
 '$ 2.7bn',
 '$ 2.9bn',
 '$ 20',
 '$ 200,000',
 '$ 200m',
 '$ 206bn',
 '$ 20bn',
 '$ 21m',
 '$ 23bn',
 '$ 24.9m',
 '$ 24m',
 '$ 25.4bn',
 '$ 250',
 '$ 250,000',
 '$ 250m',
 '$ 25m',
 '$ 27.5bn',
 '$ 28.3bn',
 '$ 280bn',
 '$ 3.1bn',
 '$ 3.5b

3.2.Transformation TF-IDF


In [ ]:
bow=CountVectorizer(tokenizer=segmenter_texte ,ngram_range=(1,2),min_df=2, max_df=5)
bow.fit(X_train)
X_train_bow=bow.transform(X_train)
X_train_bow

/usr/local/lib/python3.6/dist-packages/sklearn/feature_extraction/text.py:507: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn("The parameter 'token_pattern' will not be used"


<1557x54368 sparse matrix of type '<class 'numpy.int64'>'
	with 137103 stored elements in Compressed Sparse Row format>

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
import pandas as pd
tfidf=TfidfVectorizer(min_df=2,max_df=5, ngram_range=(1,2))
X_train_tfidf=tfidf.fit_transform(X_train)
vocabulaire=tfidf.get_feature_names()
pd.DataFrame(X_train_tfidf.todense(), columns=tfidf.get_feature_names())


00  000 12  000 293  ...  zvonareva  zvyagintsev  zvyagintsev the
0     0.0     0.0      0.0  ...        0.0          0.0              0.0
1     0.0     0.0      0.0  ...        0.0          0.0              0.0
2     0.0     0.0      0.0  ...        0.0          0.0              0.0
3     0.0     0.0      0.0  ...        0.0          0.0              0.0
4     0.0     0.0      0.0  ...        0.0          0.0              0.0
...   ...     ...      ...  ...        ...          ...              ...
1552  0.0     0.0      0.0  ...        0.0          0.0              0.0
1553  0.0     0.0      0.0  ...        0.0          0.0              0.0
1554  0.0     0.0      0.0  ...        0.0          0.0              0.0
1555  0.0     0.0      0.0  ...        0.0          0.0              0.0
1556  0.0     0.0      0.0  ...        0.0          0.0              0.0

[1557 rows x 66691 columns]

In [ ]:
bow=CountVectorizer(tokenizer=segmenter_texte ,ngram_range=(1,2),min_df=2, max_df=5)
bow.fit(X_train)
vocabulaire=bow.get_feature_names()

/usr/local/lib/python3.6/dist-packages/sklearn/feature_extraction/text.py:507: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn("The parameter 'token_pattern' will not be used"


In [ ]:
from sklearn.feature_extraction.text import TfidfTransformer
tfidf=TfidfTransformer(use_idf=True, smooth_idf=False)
tfidf.fit(X_train_bow)
pd.DataFrame(tfidf.idf_.reshape(1,-1),columns=vocabulaire,index=['poids IDF'])


! ''       ! ,   ! along  ...  £95.8m )       £9m  £9m bring
poids IDF  6.741078  6.964222  7.657369  ...  7.657369  6.741078   7.657369

[1 rows x 54368 columns]

In [ ]:
tfidf=TfidfTransformer(use_idf=True, smooth_idf=True)
tfidf.fit(X_train_bow)
pd.DataFrame(tfidf.idf_.reshape(1,-1),columns=vocabulaire,index=['poids IDF'])

! ''      ! ,   ! along  ...  £95.8m )       £9m  £9m bring
poids IDF  6.559399  6.74172  7.252546  ...  7.252546  6.559399   7.252546

[1 rows x 54368 columns]

# Etape 4 : Modèle d’apprentissage automatique (Modeling)


In [ ]:
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline
import numpy as np


In [ ]:
from nltk.corpus import stopwords
import nltk
from sklearn.feature_extraction.text import CountVectorizer




In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
import pandas as pd
from sklearn.feature_extraction.text import TfidfTransformer




In [ ]:
from sklearn import svm


In [ ]:
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline
import numpy as np

In [ ]:
pipeline1 = [("vect", CountVectorizer()), 
        ("tfidf", TfidfTransformer()),
        ("svm", SVC())]
pipe = Pipeline(pipeline1)
pipe.named_steps
kernel="rbf"

In [ ]:

params = {
    'vect__min_df': np.linspace(0.005, 0.05, 5),
    'vect__ngram_range': ((1, 1),(1, 2)), 
    'vect__decode_error': ('ignore','UTF-8'),
    'tfidf__use_idf': (True, False),
    'svm__C':  [0.1, 1, 10],
    # 'svm__gamma': [0.1, 1, 10],
    # 'svm__kernel': [0.1, 1, 10]
}


In [ ]:

gs1=GridSearchCV(pipe,param_grid=params,cv=5,n_jobs=-1)
gs1=gs1.fit(X_train, y_train)
gs1.best_params_

{'svm__C': 1,
 'tfidf__use_idf': True,
 'vect__decode_error': 'ignore',
 'vect__min_df': 0.005,
 'vect__ngram_range': (1, 2)}

In [ ]:
from sklearn.naive_bayes import MultinomialNB
pipeline2 = [("vect", CountVectorizer()), 
        ("tfidf", TfidfTransformer()),
        ("multinomialNB", MultinomialNB(alpha = 1.0))]
pipe = Pipeline(pipeline2)
pipe.named_steps

{'multinomialNB': MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True),
 'tfidf': TfidfTransformer(norm='l2', smooth_idf=True, sublinear_tf=False, use_idf=True),
 'vect': CountVectorizer(analyzer='word', binary=False, decode_error='strict',
                 dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                 lowercase=True, max_df=1.0, max_features=None, min_df=1,
                 ngram_range=(1, 1), preprocessor=None, stop_words=None,
                 strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                 tokenizer=None, vocabulary=None)}

In [ ]:
params = {
    'vect__min_df': np.linspace(0.005, 0.05, 5),
    'vect__ngram_range': ((1, 1),(1, 2)), 
    'vect__decode_error': ('ignore','UTF-8'),
    'tfidf__use_idf': (True, False),
    'multinomialNB__fit_prior' : (False,True)
}

In [ ]:
gs2=GridSearchCV(pipe,param_grid=params,cv=5,n_jobs=-1)
gs2=gs2.fit(X_train, y_train)
gs2.best_params_

{'multinomialNB__fit_prior': False,
 'tfidf__use_idf': True,
 'vect__decode_error': 'ignore',
 'vect__min_df': 0.01625,
 'vect__ngram_range': (1, 1)}

# Etape 5 : Evaluation (Evaluation)

In [ ]:
y_test_predicted1=gs1.predict(X_test)
y_test_predicted1

array([2, 2, 1, 1, 1, 5, 5, 2, 4, 5, 1, 5, 1, 3, 2, 1, 4, 2, 2, 3, 4, 4,
       3, 1, 1, 5, 1, 4, 3, 1, 5, 3, 1, 2, 4, 3, 3, 1, 2, 4, 1, 5, 5, 5,
       5, 3, 4, 4, 2, 2, 4, 5, 1, 1, 4, 4, 2, 4, 2, 4, 1, 5, 2, 1, 1, 4,
       4, 1, 3, 4, 5, 4, 3, 2, 2, 2, 4, 4, 1, 1, 5, 3, 4, 4, 4, 5, 4, 5,
       1, 5, 4, 4, 2, 3, 4, 5, 3, 3, 1, 4, 4, 1, 3, 3, 5, 5, 4, 2, 4, 2,
       1, 2, 3, 2, 3, 5, 4, 3, 3, 1, 3, 1, 2, 4, 1, 1, 1, 2, 1, 3, 5, 2,
       4, 4, 1, 5, 4, 4, 1, 2, 4, 1, 5, 1, 4, 1, 3, 4, 5, 1, 1, 2, 4, 4,
       5, 1, 2, 5, 3, 5, 1, 2, 1, 1, 2, 4, 1, 5, 1, 4, 4, 2, 2, 1, 2, 4,
       2, 5, 1, 5, 5, 4, 4, 1, 2, 5, 4, 1, 1, 4, 3, 4, 5, 1, 1, 1, 5, 2,
       1, 5, 2, 1, 4, 3, 5, 3, 1, 1, 5, 1, 2, 2, 5, 3, 3, 3, 1, 2, 2, 5,
       1, 4, 2, 1, 2, 5, 3, 1, 4, 3, 5, 4, 5, 3, 5, 1, 1, 1, 2, 1, 3, 2,
       1, 1, 2, 2, 5, 4, 1, 4, 1, 5, 4, 2, 1, 1, 1, 4, 4, 3, 3, 3, 1, 5,
       3, 5, 3, 3, 1, 1, 2, 2, 2, 2, 1, 3, 4, 5, 3, 2, 2, 2, 3, 4, 3, 4,
       2, 1, 1, 5, 2, 5, 1, 4, 2, 4, 2, 2, 4, 4, 1,

In [ ]:
y_test_predicted2=gs2.predict(X_test)
y_test_predicted2

array([2, 2, 1, 1, 1, 5, 5, 2, 4, 5, 1, 5, 1, 3, 2, 1, 4, 5, 2, 3, 4, 4,
       3, 1, 5, 2, 1, 4, 3, 1, 5, 1, 1, 2, 4, 3, 3, 1, 2, 4, 1, 5, 5, 5,
       5, 3, 4, 4, 2, 2, 4, 5, 1, 1, 4, 4, 2, 4, 2, 4, 1, 5, 2, 1, 1, 4,
       4, 1, 3, 4, 5, 4, 3, 2, 2, 2, 4, 4, 1, 1, 5, 3, 4, 4, 4, 5, 4, 5,
       1, 5, 4, 4, 2, 3, 4, 5, 3, 3, 1, 4, 4, 1, 3, 3, 5, 5, 4, 2, 4, 2,
       1, 2, 3, 2, 3, 5, 4, 3, 3, 1, 3, 1, 2, 4, 1, 1, 1, 2, 1, 3, 5, 2,
       4, 4, 1, 5, 4, 4, 1, 2, 4, 1, 5, 1, 4, 1, 3, 4, 5, 1, 1, 2, 4, 4,
       5, 1, 2, 5, 3, 5, 1, 2, 1, 1, 2, 4, 1, 5, 1, 4, 4, 4, 2, 1, 2, 4,
       2, 5, 1, 5, 5, 4, 4, 1, 5, 5, 4, 1, 1, 4, 3, 4, 5, 1, 1, 1, 5, 2,
       1, 5, 2, 1, 4, 3, 5, 3, 1, 1, 5, 1, 2, 2, 5, 3, 3, 3, 1, 2, 2, 5,
       5, 4, 2, 1, 2, 5, 3, 1, 4, 3, 5, 4, 5, 3, 5, 1, 1, 1, 2, 1, 3, 4,
       1, 5, 2, 2, 5, 4, 1, 4, 1, 5, 4, 2, 1, 1, 1, 4, 4, 3, 3, 3, 1, 5,
       3, 5, 3, 3, 1, 1, 5, 2, 2, 2, 1, 3, 4, 5, 3, 2, 2, 2, 3, 4, 3, 4,
       2, 1, 1, 5, 2, 5, 5, 4, 2, 4, 2, 2, 4, 4, 1,

In [ ]:
from sklearn.metrics import accuracy_score,confusion_matrix
print(accuracy_score(y_test,y_test_predicted1))
print(confusion_matrix(y_test,y_test_predicted1))


0.9760479041916168
[[160   0   2   0   1]
 [  1 114   3   0   2]
 [  3   0 115   0   1]
 [  0   0   0 159   0]
 [  2   1   0   0 104]]


In [ ]:
print(accuracy_score(y_test,y_test_predicted2))
print(confusion_matrix(y_test,y_test_predicted2))

0.9580838323353293
[[156   1   3   0   3]
 [  0 110   3   2   5]
 [  5   0 113   0   1]
 [  0   0   0 159   0]
 [  2   1   2   0 102]]


In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_test_predicted1))

              precision    recall  f1-score   support

           1       0.96      0.98      0.97       163
           2       0.99      0.95      0.97       120
           3       0.96      0.97      0.96       119
           4       1.00      1.00      1.00       159
           5       0.96      0.97      0.97       107

    accuracy                           0.98       668
   macro avg       0.98      0.97      0.97       668
weighted avg       0.98      0.98      0.98       668



In [ ]:
print(classification_report(y_test,y_test_predicted2))

              precision    recall  f1-score   support

           1       0.96      0.96      0.96       163
           2       0.98      0.92      0.95       120
           3       0.93      0.95      0.94       119
           4       0.99      1.00      0.99       159
           5       0.92      0.95      0.94       107

    accuracy                           0.96       668
   macro avg       0.96      0.96      0.96       668
weighted avg       0.96      0.96      0.96       668



#Etape 6 : Dépoilement (Deployment)


In [ ]:
import pickle

In [ ]:
gs1=GridSearchCV(pipe,param_grid=params,cv=5,n_jobs=-1)
gs1=gs1.fit(X_train, y_train)
gs1.best_params_

{'multinomialNB__fit_prior': False,
 'tfidf__use_idf': True,
 'vect__decode_error': 'ignore',
 'vect__min_df': 0.01625,
 'vect__ngram_range': (1, 1)}

In [ ]:
with open('modele3.dat','w') as f:
    pickle.dumps(gs1)
    # f.write(s)
f.close()

In [ ]:
 with open('modele3.dat', 'r') as f:
    f.read()
f.close()

In [ ]:
gs1 = None

In [ ]:
with open('modele3.dat', 'rb') as f:
    gs1 = pickle.load(f)
    gs1.best_params_
f.close()

EOFError: ignored

In [ ]:
gs1.fit(X_new,Y_new)